File: examples/chapter17/e03_split_shuffle/python/main.py
"""Example 17.3 - Splitting and Shuffling (PyTorch reference + guarantee test).

A seeded 16/4 train/validation split, and a DataLoader that reshuffles the
training set every epoch. PyTorch's random numbers are not Burn's, so the
orders differ from rust.rs; the test checks the same guarantees instead.

Run:   python main.py
Test:  pytest main.py
"""

import torch
from torch.utils.data import DataLoader, random_split


def build():
    train, val = random_split(range(20), [16, 4], generator=torch.Generator().manual_seed(42))
    train_ids, val_ids = list(train), list(val)

    def make_loader():
        return DataLoader(train_ids, batch_size=4, shuffle=True,
                          generator=torch.Generator().manual_seed(7))

    loader = make_loader()
    epoch1 = [int(i) for b in loader for i in b]
    epoch2 = [int(i) for b in loader for i in b]
    again = [int(i) for b in make_loader() for i in b]
    return train_ids, val_ids, epoch1, epoch2, again


def main():
    train, val, epoch1, epoch2, again = build()
    print(f"train ({len(train)}) =", train)
    print(f"val   ({len(val)})  =", val)
    print("epoch 1    =", epoch1)
    print("epoch 2    =", epoch2)
    print("epoch 1 of a rebuilt loader, same seed =", "identical" if again == epoch1 else "DIFFERENT")
    print("each epoch sees every training item exactly once:",
          sorted(epoch1) == sorted(train) and sorted(epoch2) == sorted(train))


def test_same_guarantees_as_burn():
    train, val, epoch1, epoch2, again = build()
    assert (len(train), len(val)) == (16, 4)
    assert sorted(train + val) == list(range(20))
    assert sorted(epoch1) == sorted(train) and sorted(epoch2) == sorted(train)
    assert epoch1 != epoch2
    assert epoch1 == again


if __name__ == "__main__":
    main()

File: examples/chapter17/e03_split_shuffle/rust/src/main.rs
use std::sync::Arc;

use burn::backend::NdArray;
use burn::data::dataloader::DataLoaderBuilder;
use burn::data::dataloader::batcher::Batcher;
use burn::data::dataset::transform::{PartialDataset, ShuffledDataset};
use burn::data::dataset::{Dataset, InMemDataset};
use burn::tensor::backend::Backend;

type MyBackend = NdArray;

// Items are just their own index, 0..20, so every order is easy to read.
#[derive(Clone, Default)]
struct Ids;

impl<B: Backend> Batcher<B, u32, Vec<u32>> for Ids {
    fn batch(&self, items: Vec<u32>, _device: &B::Device) -> Vec<u32> {
        items
    }
}

fn is_permutation(mut seen: Vec<u32>, of: &[u32]) -> bool {
    let mut want = of.to_vec();
    seen.sort();
    want.sort();
    seen == want
}

/// Returns (train ids, val ids, epoch-1 order, epoch-2 order, epoch-1 order of a rebuilt loader).
fn build() -> (Vec<u32>, Vec<u32>, Vec<u32>, Vec<u32>, Vec<u32>) {
    let all = InMemDataset::new((0..20u32).collect::<Vec<_>>());

    // 1. Shuffle ONCE with a fixed seed, then cut: 16 for training, 4 for validation.
    //    Shuffle before splitting, or the validation set is just "the last rows of the file".
    let shuffled = Arc::new(ShuffledDataset::new(all, 42));
    let train = PartialDataset::new(shuffled.clone(), 0, 16);
    let val = PartialDataset::new(shuffled, 16, 20);
    let train_ids: Vec<u32> = train.iter().collect();
    let val_ids: Vec<u32> = val.iter().collect();

    // 2. A loader that reshuffles the training set EVERY epoch, from one seed.
    let make_loader = || {
        DataLoaderBuilder::<MyBackend, u32, Vec<u32>>::new(Ids)
            .batch_size(4)
            .shuffle(7)
            .build(InMemDataset::new(train_ids.clone()))
    };
    let loader = make_loader();
    let epoch1: Vec<u32> = loader.iter().flatten().collect();
    let epoch2: Vec<u32> = loader.iter().flatten().collect();
    let again: Vec<u32> = make_loader().iter().flatten().collect(); // same seed, fresh loader

    (train_ids, val_ids, epoch1, epoch2, again)
}

fn main() {
    let (train, val, epoch1, epoch2, again) = build();
    println!("train ({}) = {train:?}", train.len());
    println!("val   ({})  = {val:?}", val.len());
    println!("epoch 1    = {epoch1:?}");
    println!("epoch 2    = {epoch2:?}");
    println!(
        "epoch 1 of a rebuilt loader, same seed = {}",
        if again == epoch1 {
            "identical"
        } else {
            "DIFFERENT"
        }
    );
    println!(
        "each epoch sees every training item exactly once: {}",
        is_permutation(epoch1.clone(), &train) && is_permutation(epoch2.clone(), &train)
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn same_guarantees_as_pytorch() {
        // The RNGs differ, so the ORDERS differ from python.py. The guarantees must not.
        let (train, val, epoch1, epoch2, again) = build();
        assert_eq!((train.len(), val.len()), (16, 4));
        let mut all: Vec<u32> = train.iter().chain(&val).copied().collect();
        all.sort();
        assert_eq!(all, (0..20).collect::<Vec<u32>>()); // disjoint, and nothing lost
        assert!(is_permutation(epoch1.clone(), &train));
        assert!(is_permutation(epoch2.clone(), &train));
        assert_ne!(epoch1, epoch2); // a new order every epoch
        assert_eq!(epoch1, again); //  ...but reproducible from the seed
    }
}
